src/backend/cdb/cdbmutate.c

```diff
@@ -788,6 +793,82 @@ apply_motion_mutator(Node *node, ApplyMotionState *context)
 			return plan_tree_mutator(node, apply_motion_mutator, context);
 	}
 
+	/*
+	 * For UPDATE/DELETE, we check if there's any motions before scan in the
+	 * same subtree for the table we're going to modify. If we encounter the
+	 * scan before any motions, then we can elide unnecessary Explicit
+	 * Redistribute Motion.
+	 */
+	if (IsA(node, ModifyTable))
+	{
+		ModifyTable *mt = (ModifyTable *) node;
+
+		if (mt->operation == CMD_UPDATE || mt->operation == CMD_DELETE)
+		{
+			/*
+			 * Sanity check, since we don't allow multiple ModifyTable nodes.
+			 */
+			Assert(!context->mtIsChecking);
+			Assert(context->mtResultRelations == NIL);
+			Assert(context->nMotionsAbove == 0);
+
+			/*
+			 * When UPDATE/DELETE occurs on a partitioned table, or a table that
+			 * is a part of inheritance tree, ModifyTable node will have more
+			 * than one relation in resultRelations.
+			 *
+			 * Remember resulting relations' indexes to compare them later.
+			 */
+			context->mtIsChecking = true;
+			context->mtResultRelations = mt->resultRelations;
+		}
+	}
+	else if (context->mtIsChecking)
+	{
+		/* 
+		 * Remember if we are descending into a motion node.
+		 */
+		if (IsA(node, Motion))
+			context->nMotionsAbove++;
+		else
+		{
+			/*
+			 * If this is a scan and it's scanrelid matches ModifyTable's relid,
+			 * we need to check if there were any motions above.
+			 *
+			 * These are scan nodes that can be used to perform distributed
+			 * UPDATE/DELETE on the relation they scan, possibly with motions
+			 * above them. This list needs to be updated for other nodes if they
+			 * are changed to support DML execution on segments.
+			 */
+			switch (nodeTag(node))
+			{
+				case T_SeqScan:
+				case T_DynamicSeqScan:
+				case T_IndexScan:
+				case T_DynamicIndexScan:
+				case T_IndexOnlyScan:
+				case T_BitmapIndexScan:
+				case T_DynamicBitmapIndexScan:
+				case T_BitmapHeapScan:
+				case T_DynamicBitmapHeapScan:
+				case T_TidScan:
+					if (list_member_int(context->mtResultRelations,
+										((Scan *) node)->scanrelid))
+					{
+						/*
+						 * Freeze the motion counter. Also, we don't
```

Comment:
Where is counter here?

Reply:
`context->nMotionsAbove` is the counter.

>> Any motions found underneath InitPlans are ignored.
>
> where is it in the code?

I removed this part from the description, since it's stale now. Now, InitPlans are handled because of the fact that different aliases on the same table have different range table indexes.